alacritty_terminal/src/tty/windows/child.rs

```diff
@@ -0,0 +1,128 @@
+use winapi::um::winbase::{RegisterWaitForSingleObject, UnregisterWait, INFINITE};
+use winapi::um::winnt::{WT_EXECUTEINWAITTHREAD, WT_EXECUTEONLYONCE};
+
+use crate::tty::ChildEvent;
+
+/// WinAPI callback to run when child process exits.
+extern "system" fn child_exit_callback(ctx: PVOID, timed_out: BOOLEAN) {
+    if timed_out != 0 {
+        return;
+    }
+
+    let event_tx: Box<_> = unsafe { Box::from_raw(ctx as *mut Sender<ChildEvent>) };
+    let _ = event_tx.send(ChildEvent::Exited);
+}
+
+pub struct ChildExitWatcher {
+    wait_handle: AtomicPtr<c_void>,
+    event_rx: Receiver<ChildEvent>,
+}
+
+impl ChildExitWatcher {
+    pub fn new(child_handle: HANDLE) -> Result<ChildExitWatcher, Error> {
+        let (event_tx, event_rx) = channel::<ChildEvent>();
+
+        let mut wait_handle: HANDLE = 0 as HANDLE;
+        let sender_ref = Box::new(event_tx);
+
+        let success = unsafe {
+            RegisterWaitForSingleObject(
+                &mut wait_handle,
+                child_handle,
+                Some(child_exit_callback),
+                Box::into_raw(sender_ref) as PVOID,
+                INFINITE,
+                WT_EXECUTEINWAITTHREAD | WT_EXECUTEONLYONCE,
+            )
+        };
+
+        if success == 0 {
+            Err(Error::last_os_error())
+        } else {
+            Ok(ChildExitWatcher { wait_handle: AtomicPtr::from(wait_handle), event_rx })
+        }
+    }
+
+    pub fn event_rx(&self) -> &Receiver<ChildEvent> {
+        &self.event_rx
+    }
+}
+
+impl Drop for ChildExitWatcher {
+    fn drop(&mut self) {
+        unsafe {
+            UnregisterWait(self.wait_handle.load(Ordering::Relaxed));
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::os::windows::io::AsRawHandle;
+    use std::process::Command;
+    use std::time::Duration;
+
+    use mio::{Events, Poll, PollOpt, Ready, Token};
+
+    use super::*;
+
+    #[test]
+    pub fn event_is_emitted_when_child_exits() {
+        const WAIT_TIMEOUT: Duration = Duration::from_millis(200);
+
+        let mut child = Command::new("cmd.exe").spawn().unwrap();
+        let child_exit_watcher = ChildExitWatcher::new(child.as_raw_handle()).unwrap();
+
+        let poll = Poll::new().unwrap();
+        let mut events = Events::with_capacity(8);
+        let child_events_token = Token::from(0usize);
+
+        poll.register(
+            child_exit_watcher.event_rx(),
+            child_events_token,
+            Ready::readable(),
+            PollOpt::oneshot(),
+        )
+        .unwrap();
+
+        child.kill().unwrap();
+
+        // Poll for the event or fail with timeout if nothing has been sent
+        poll.poll(&mut events, Some(WAIT_TIMEOUT)).unwrap();
+        for event in events.iter() {
+            if event.token() == child_events_token {
+                // Verify that at least one `ChildEvent::Exited` was received
+                loop {
+                    let result = child_exit_watcher.event_rx().try_recv();
```

Comment:
This is going to produce an extremely hot loop just going at it over and over again, which is not ideal.

We should just `event_rx().recv()` on this.

Reply:
Don't think that would be a hot loop. What it will do is take all the events queued on the received and discard all of them until it either gets `Ok(ChildEvent::Exited)` ending in success or `Err(TryRecvError::Empty)` breaking the loop and failing.

Reply:
Yeah, but at that point why are we using `try_recv` instead of iterating?

Reply:
Not sure what should I iterate on? Only way too verify that the right event type was send from the watcher thread is to receive them on the receiver side.

Reply:
With the iterator I was talking about the `iter` method which is provided on a `Receiver`, which allows iterating until `try_recv` would block.

However I think that just taking next and asserting that is the better choice and much simpler to understand, assuming we don't send any other events before that yet.